pageserver/src/layered_repository/inmemory_layer.rs

```diff
@@ -427,8 +484,10 @@ impl InMemoryLayer {
         assert!(lsn >= inner.latest_lsn);
         inner.latest_lsn = lsn;
 
-        let old = inner.page_versions.append_or_update_last(blknum, lsn, pv)?;
-
+        // Write the page version to the file, and remember its offset in 'page_versions'
+        let off = inner.write_pv(&pv)?;
+        let vec_map = inner.page_versions.entry(blknum).or_default();
+        let old = vec_map.append_or_update_last(lsn, off).unwrap().0;
```

Comment:
> This is complex enough to be a big distraction here, especially the "remember its offset" part. All things that need remembering should have as small scope as possible. The scope of this `off` is way larger than it needs to be.

I added braces around the block (and the later one), to make it more clear that it's one operation, and to limit the scope of 'off'.

> Maybe this needs a helper function, ...

Hmm, perhaps. However, PR #1267 moves the code that zero-fills the "gap" to layered_repository.rs, so the second block will be gone with that. So I think this is fine.

> ... ideally one that returns Err instead of silently updating last.

Yeah, that would be nice. Currently, I think there are some corner cases where we call put_page_version() twice for the same block and LSN. One such case is if a WAL record extends a relation by two blocks, but the blocks are listed in the WAL record in backwards order. That results in two calls to put_page_version(), for example, first for block 2, and then for 1. The first call, for block 2, would implicitly store an all-zeros page for block 1, thanks to the "gap" handling code. But then the second call would overwrite it with the real page image.

I'm not 100% sure if that can happen in reality. We should investigate. And if it can happen, I think we should modify walingest.rs so that makes the put_page_version calls in right order to avoid it. But that's a separate story, this PR doesn't try to change that.